src/core/kmod-setup.c

```diff
@@ -80,6 +81,10 @@ static bool has_virtio_rng(void) {
 
         return r > 0;
 }
+
+static bool in_qemu(void) {
+        return IN_SET(detect_vm(), VIRTUALIZATION_KVM, VIRTUALIZATION_QEMU);
```

Comment:
mmh are we sure about `VIRTUALIZATION_KVM`? There are other hypervisors that run on KVM, eg: https://github.com/cloud-hypervisor/cloud-hypervisor would those be matched too here?

Reply:
docs clearly say "kvm" means "qemu+kvm".

Quoting systemd-detect-vrit man page:

> "Linux KVM kernel virtual machine, in combination with QEMU. Not used for other virtualizers using the KVM interfaces, such as Oracle VirtualBox or Amazon EC2 Nitro, see below."

Reply:
(I mean, of course, in retrospect we should nevre have added the explicit "kvm" enum value ere and just make it "qemu", but back then it wasn't clear that kvm would be used by a number of different hypervisors one day)

Reply:
Well, the KVM CPUID stuff predates systemd-detect-virt, I think. Running under cloud-hypervisor we indeed detect kvm:

```
$ ./cloud-hypervisor-static --kernel ./hypervisor-fw --disk path=/tmp/focal-server-cloudimg-amd64.raw --cpus boot=2 --memory size=2048M
cloud-hypervisor: 5.568027101s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x87
cloud-hypervisor: 5.606556807s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x2f9
cloud-hypervisor: 5.606595571s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x2f9
cloud-hypervisor: 5.606628522s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x2f9
cloud-hypervisor: 5.606670021s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x2f9
cloud-hypervisor: 5.60670979s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x2f9
cloud-hypervisor: 5.606730758s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x2f9
cloud-hypervisor: 5.60681496s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x3e9
cloud-hypervisor: 5.606844019s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x3e9
cloud-hypervisor: 5.606869118s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x3e9
cloud-hypervisor: 5.606894391s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x3e9
cloud-hypervisor: 5.6069225s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x3e9
cloud-hypervisor: 5.606949647s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x3e9
cloud-hypervisor: 5.607026592s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x2e9
cloud-hypervisor: 5.607053484s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x2e9
cloud-hypervisor: 5.60707703s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x2e9
cloud-hypervisor: 5.607101702s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x2e9
cloud-hypervisor: 5.607126966s: <vcpu0> WARN:vmm/src/vm.rs:470 -- Guest PIO read to unregistered address 0x2e9
cloud-hypervisor: 5.607152985s: <vcpu0> WARN:vmm/src/vm.rs:495 -- Guest PIO write to unregistered address 0x2e9

Ubuntu 20.04.4 LTS ubuntu hvc0

ubuntu login: root
Password: 
Welcome to Ubuntu 20.04.4 LTS (GNU/Linux 5.4.0-109-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

  System information as of Mon May  2 15:02:13 UTC 2022

  System load: 0.23              Memory usage: 9%   Processes:       122
  Usage of /:  67.5% of 1.96GB   Swap usage:   0%   Users logged in: 0

1 update can be applied immediately.
To see these additional updates run: apt list --upgradable



The programs included with the Ubuntu system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Ubuntu comes with ABSOLUTELY NO WARRANTY, to the extent permitted by
applicable law.

root@ubuntu:~# SYSTEMD_LOG_LEVEL=debug systemd-detect-virt 
Found container virtualization none.
No virtualization found in DMI
Virtualization found, CPUID=KVMKVMKVM
Found VM virtualization kvm
kvm
```

So this would break things, no?

Reply:
well, i think if there is a VM we currently don't detect correctly, we should fix that.

Note that various other implementations that use kvm actually pass a different CPUID than "KVMKVMKVM". I think it's kinda sloppy if "could-hypervisor" uses KVMKVMKVM for that...

But anyway, I think we should stick to our definition of "kvm", and add detection entries for other VM hypervisors as necessary. We followed this rule so far, I see no reason to depart from it now.